pkg/snapstore/swift_snapstore.go

```diff
@@ -483,13 +484,47 @@ func (s *SwiftSnapStore) List() (brtypes.SnapList, error) {
 
 	sort.Sort(snapList)
 	return snapList, nil
+}
+
+func (s *SwiftSnapStore) getSnapshotChunks(snapshot brtypes.Snapshot) (brtypes.SnapList, error) {
+	snaps, err := s.List()
+	if err != nil {
+		return nil, err
+	}
 
+	var chunkList brtypes.SnapList
+	for _, snap := range snaps {
+		if snap.IsChunk {
+			chunkParentSnapPath, _ := path.Split(path.Join(snap.Prefix, snap.SnapDir, snap.SnapName))
+			if strings.TrimSuffix(chunkParentSnapPath, "/") == path.Join(snapshot.Prefix, snapshot.SnapDir, snapshot.SnapName) {
+				chunkList = append(chunkList, snap)
+			}
+		}
+	}
+	return chunkList, nil
 }
 
-// Delete should delete the snapshot file from store
+// Delete deletes the objects related to the DLO (dynamic large object) from the store.
+// This includes the manifest object as well as the segment objects, as
+// described in https://docs.openstack.org/swift/latest/overview_large_objects.html
 func (s *SwiftSnapStore) Delete(snap brtypes.Snapshot) error {
-	result := objects.Delete(s.client, s.bucket, path.Join(snap.Prefix, snap.SnapDir, snap.SnapName), nil)
-	return result.Err
+	chunks, err := s.getSnapshotChunks(snap)
+	if err != nil {
+		return err
+	}
+
+	var chunkObjectNames []string
+	for _, chunk := range chunks {
+		chunkObjectNames = append(chunkObjectNames, path.Join(chunk.Prefix, chunk.SnapDir, chunk.SnapName))
+	}
+
+	manifestObjectDeleteResult := objects.Delete(s.client, s.bucket, path.Join(snap.Prefix, snap.SnapDir, snap.SnapName), nil)
+	if manifestObjectDeleteResult.Err != nil {
+		return manifestObjectDeleteResult.Err
+	}
```

Comment:
What if after the manifest object is deleted without any problem but then it gets the error deleting the chunk object in bulk in rare network issues. These chunks will be in the store forever because their manifest is deleted, so we won't get to this place ever again to delete those orphan chunks.

So, for deleting chunks like these, can we allow them to be deleted in the GC.

Or instead of dealing with the chunks here, can we delete all the chunks from GC with existing logic if they are out of our threshold holding time i.e 4 weeks. Then we can reduce the workload on snapshotter and make all the work through GC. Just a suggestion because when I raised the PR, I was trying to minimise workload on snapshotter.

Reply:
@anveshreddy18 this is a good point. Based on our out-of-band discussion yesterday, I have now changed the order of deletion of the objects to solve this problem. The chunks will be deleted first, followed by the manifest object. This way, if there is an error in deletion of the chunks, then the `Delete()` operation can be retried in the next run of garbage collector, since the manifest object still exists (the `List()` call simply lists all non-chunk objects, aka the manifest objects, which are considered to be the snapshot objects by any user of snapstore). This will allow for re-attempting of deletion of the remaining chunk objects, as returned by `getSnapshotChunks()` function. 

If chunk object deletion succeeds, but manifest object deletion fails, then the same logic holds - the Delete() operation will be retried in the next iteration of GC, which will then re-attempt to delete the manifest object.

This way, we don't need to have any additional handling of openstack chunk objects in garbage collector. This way the chunks are deleted along with the manifest object, and not retained for 4 weeks. This helps with cost savings, since buckets are charged as per the amount of data stored in it doe a given duration, like GB/month.

Reply:
Thanks Shreyas for addressing my concern
> I have now changed the order of deletion of the objects to solve this problem.

Great! This will work. But I just have a small doubt.
Consider the scenario: 

- After all the chunks are deleted ( could be in multiple cycles of GC )
- Manifest object failed to delete for some reason in the latest cycle of GC
- In the next GC cycle, when it attempts to delete the chunks again, it gets empty list and hence `chunkObjectNames` is an empty slice of strings. 

Now I'm not sure what happens when the line `chunkObjectsDeleteResult := objects.BulkDelete(s.client, s.bucket, chunkObjectNames)` is executed as `chunkObjectNames` list is empty, I'm hoping `chunkObjectsDeleteResult.Err` would be nil. I tried searching for it, but couldn't find much documentation on that.  PTAL about this once.  If you are confident that there won't be any error from your knowledge of dealing with cloud providers, then all good.

Reply:
Good point @anveshreddy18 . I'll add a check for the length of `chunks` returned, and call BulkDelete() based on that.

Reply:
Done